scipy/optimize/_qap.py

```diff
@@ -0,0 +1,418 @@
+
+    >>> n = cost.shape[0]
+    >>> P = np.zeros((n, n))
+    >>> P[np.arange(n), res['col_ind']] = 1
+    >>> score = int(np.trace(cost.T @ P @ dist @ P.T))
+    >>> print(score)
+    3260
+
+    As you can see, the value here matches res['score'] reported above.
+    Alternatively, to avoid constructing the permutation matrix, one can also
+    perform the following calculation.
+
+    >>> score = np.trace(cost.T @ dist[np.ix_(res['col_ind'], res['col_ind'])])
+    >>> print(score)
+    3260
+
+    Here, we are simply permuting the distance matrix.
+
+    """
+
+    cost_matrix = np.asarray(cost_matrix)
+    dist_matrix = np.asarray(dist_matrix)
+
+    if seed is None:
+        seed = np.array([[], []]).T
+    seed = np.asarray(seed)
+
+    # ValueError check
+    msg = None
+    if cost_matrix.shape[0] != dist_matrix.shape[0]:
+        msg = "Adjacency matrices must be of equal size"
+    elif (
+        cost_matrix.shape[0] != cost_matrix.shape[1]
+        or dist_matrix.shape[0] != dist_matrix.shape[1]
+    ):
+        msg = "Adjacency matrix entries must be square"
+    elif not (cost_matrix >= 0).all() or not (dist_matrix >= 0).all():
+        msg = "Adjacency matrix entries must be greater than or equal to zero"
+    elif seed.shape[0] > cost_matrix.shape[0]:
+        msg = "There cannot be more seeds than there are nodes"
+    elif seed.shape[1] != 2:
+        msg = "Seed array entry must have two columns"
+    elif not (seed >= 0).all():
+        msg = "Seed array entries must be greater than or equal to zero"
+    elif not (seed <= (cost_matrix.shape[0] - 1)).all():
+        msg = "Seed array entries must be less than the number of nodes"
+    elif not len(set(seed[:, 0])) == len(seed[:, 0]) or not \
+            len(set(seed[:, 1])) == len(seed[:, 1]):
+        msg = "Seed column entries must be unique"
+    elif init_method not in {'barycenter', 'rand'}:
+        msg = "Invalid 'init_method' parameter string"
+    if msg is not None:
+        raise ValueError(msg)
+
+    # TypeError check
+    if type(n_init) is not int or n_init <= 0:
+        msg = "'n_init' must be a positive integer"
+    elif max_iter <= 0 or type(max_iter) is not int:
+        msg = "'max_iter' must be a positive integer"
+    elif type(shuffle_input) is not bool:
+        msg = "'shuffle_input' must be a boolean"
+    elif eps <= 0 or type(eps) is not float:
+        msg = "'eps' must be a positive float"
+    elif type(maximize) is not bool:
+        msg = "'maximize' must be a boolean"
+    if msg is not None:
+        raise TypeError(msg)
+
+    n = cost_matrix.shape[0]  # number of vertices in graphs
+    n_seeds = seed.shape[0]  # number of seeds
+    n_unseed = n - n_seeds
+
+    perm_inds = np.zeros(n)
+
+    obj_func_scalar = 1
+    if maximize:
+        obj_func_scalar = -1
+    score = obj_func_scalar * math.inf
+
+    seed_dist_c = np.setdiff1d(range(n), seed[:, 1])
+    if shuffle_input:
+        seed_dist_c = np.random.permutation(seed_dist_c)
```

Comment:
The random seed has not been set. Create a random state with `rng = np.random.RandomState(seed=seed)`, and call it with `rng.permutation`.

I also suggest renaming the current `seed` variable to `x0`, which is more in line with the existing scipy naming practices.

Reply:
I agree that maybe the name `seed` for this is confusing, but I also worry that `x0` implies they are an initial guess that gets updated iteratively (like what @asaadeldin11 specifies currently via `init_method`). It's probably important to make clear that these are "hard" seeds that will never get unmatched once they are specified.

I can look for other names in the literature, but really they are `known_matchings`, or thought of another way, the current `seeds` argument let you make this into a semi-supervised problem where having a few known matchings between elements can greatly speed up the optimization or make it more accurate (Figure 2 here https://arxiv.org/pdf/1209.0367.pdf)

Reply:
OK. I can see that `x0` would be confusing too.
How about `partial` or `partial_solution` or somesuch?

Reply:
I could see `partial_solution` making sense - do you have thoughts on whether passing in a `n_seed` x `2` array indicating the indices of matching is an intuitive way to provide this partial solution?